### PR TITLE
Add forgot/reset password feature (#1672)

### LINKED ISSUE #1672: Implement Forgot Password process in TodoTemplate
Currently, the `Forgot Password` process is not implemented in the TodoTemplate project. This process is required in a boilerplate project such as the TodoTemplate.
For now, it is required to have the base implementation of this process in the server+client, and the final UI enhancements based on new designs will be taken care of in the future.

### PR DESCRIPTION
**Pull request by mohabbati**
This closes #1672

### CODE REVIEW

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 1]
```diff
@@ -83,11 +88,49 @@ public async Task SendEmailConfirmLink(SendEmailConfirmLinkRequestDto sendConfir
         var result = await _fluentEmail
             .To(user.Email, user.DisplayName)
             .Subject(EmailStrings.ConfirmationLinkEmailSubject)
-            .UsingTemplateFromEmbedded("TodoTemplate.Api.Resources.EmailConfirmation.cshtml",
-                                    new EmailConfirmationModel
+            .UsingTemplateFromEmbedded($"TodoTemplate.Api.Resources.EmailConfirmation.cshtml",
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:91]: No need to $

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 2]
```diff
@@ -83,11 +88,49 @@ public async Task SendEmailConfirmLink(SendEmailConfirmLinkRequestDto sendConfir
...
+                                    new EmailLinksModel
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:92]: Keep using separated models for each email template. Try using inheritance if some / all properties are in common.
Note that this will gonna change and we need to add multiple properties as we move forward

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 3]
```diff
@@ -83,11 +88,49 @@ public async Task SendEmailConfirmLink(SendEmailConfirmLinkRequestDto sendConfir
...
+                                    },
+                                    assembly)
+            .SendAsync(cancellationToken);
+
+        if (!result.Successful)
+            throw new ResourceValidationException(result.ErrorMessages.ToArray());
+    }
+
+    [HttpPost("[action]")]
+    public async Task SendEmailForgotPasswordLink(SendEmailForgotPasswordLinkRequestDto sendConfirmLinkEmailRequest, CancellationToken cancellationToken)
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:105]: SendPasswordResetTokenEmail instead

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 4]
```diff
@@ -83,11 +88,49 @@ public async Task SendEmailConfirmLink(SendEmailConfirmLinkRequestDto sendConfir
...
+    {
+        var user = await _userManager.FindByEmailAsync(sendConfirmLinkEmailRequest.Email);
+
+        if (user is null)
+            throw new BadRequestException(nameof(ErrorStrings.UserNameNotFound));
+
+        var token = await _userManager.GeneratePasswordResetTokenAsync(user);
+
+        var tokenBase64 = Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(token));
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: Do we need base 64 encoding? Why we didn't the same for email confirmation?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 5]
```diff
@@ -83,11 +88,49 @@ public async Task SendEmailConfirmLink(SendEmailConfirmLinkRequestDto sendConfir
...
+
+        var resetPasswordLink = $"reset-password?email={user.Email}&token={tokenBase64}";
+
+#if BlazorServer
+        resetPasswordLink = $"{_appSettings.WebServerAddress}{resetPasswordLink}";
+#else
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:120]: The relative url would be enough, no need to #else

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/TodoTemplate.Api.csproj [hunk 6]
```diff
@@ -14,6 +14,7 @@
         <Content Condition=" '$(BlazorMode)' == 'BlazorWebAssembly' " Include="..\Web\Pages\_Layout.cshtml" Link="Pages\_Layout.cshtml" />
         <Compile Condition=" '$(BlazorMode)' == 'BlazorWebAssembly' " Include="..\Web\Services\Implementations\ServerSideAuthTokenProvider.cs" Link="Services\Implementations\ServerSideAuthTokenProvider.cs" />
         <EmbeddedResource Include="Resources\*.cshtml" />
+        <Content Remove="Resources\ResetPassword.cshtml" />
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/TodoTemplate.Api.csproj:17]: Remove this

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor [hunk 7]
```diff
@@ -58,7 +58,7 @@
                 <BitLabel>Full name</BitLabel>
                 <BitTextField @bind-Value="UserToEdit.FullName"
                           Placeholder="Full name"
-                          OnKeyUp="WrapHandled(CheckSaveButtonEnabled)" />
+                          OnChange="WrapHandled(CheckSaveButtonEnabled)" />
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor:61]: What happened here?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        if (IsLoading)
+        {
+            return;
+        }
+
+        IsLoading = true;
+        IsSubmitButtonEnabled = false;
+        ForgotPasswordMessage = null;
+
+        if (ValidateForgotPassword() is false)
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs:65]: Check ValidateForgotPassword before setting IsLoading to true, so you won't have to set it to false again

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    [Inject] public ITodoTemplateAuthenticationService TodoTemplateAuthenticationService { get; set; } = default!;
+
+    public BitMessageBarType ResetPasswordMessageType { get; set; }
+    public string? ResetPasswordMessage { get; set; }
+
+    protected override void OnInitialized()
+    {
+        base.OnInitialized();
+
+        var base64EncodedBytes = Convert.FromBase64String(Token!);
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:35]: Do we need this?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        if (IsLoading)
+        {
+            return;
+        }
+
+        IsLoading = true;
+        IsSubmitButtonEnabled = false;
+        ResetPasswordMessage = null;
+
+        if (ValidateResetPassword() is false)
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:96]: Check ValidateResetPassword before IsLoading set to true (Put it in line 91)

**Review by ysmoradi — CHANGES_REQUESTED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 2 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:92]: So, do I have to create two empty classes inherited from the EmailLinksModel and use them? @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: The problem is '+' sign in the token. It replaces with space so the token became invalid. I encode the token with base64 to solve the issue. Unlike the reset password razor page, there is no problem with the confirmation method in the Auth controller. @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor (hunk 7 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor:61]: If you copy a value and then paste it to the BitTextField using mouse it was not going to change the enable status of it. OnChange solves the issue. @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs (hunk 8 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs:65]: I changed the code as you said. Keep in mind this issue already exist on other razor pages like sign-in. @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs (hunk 9 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:35]: Asp .net core identity needs the generated token to reset password and as I said due to '+' sign problem I encode the token and now have to decode it. @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 5 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:120]: I'm not sure about that, the generated link has been sent to the user by email just 2 lines after. So, we need an absolute url. If there is another issue please explain more. @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: I think base64 encoding is overkill here, it just needs an `UrlEncoding`.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 2 above)
Comment (ysmoradi) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:92]: I'd recommend creating two separated models called "EmailConfirmationModel" & "ResetPasswordModel"
Not interested in "EmailLinksModel" at all

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (ysmoradi) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: What's the difference between the reset password token and email confirmation token?
I think both need or don't need URL encoding? @mohabbati

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 3 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:105]: it's even better to rename it to `SendResetPasswordEmail`.
it's not just the token or a link, it's a complete Email containing a link with a token :)

and also it's a lot better to rename the other method to `SendConfirmationEmail`.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: I'm pretty sure that both need the url-encoding. any dynamic data that goes to a url need this encoding, you can't be sure about the illegal characters there.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor (hunk 7 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor:61]: is it going to handle every scenario? plz confirm it with @maryamhdr and @sajadkardel 
anyway, this change is not related to this issue, so plz remove it and submit a new issue if required.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs (hunk 8 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs:65]: plz submit the issue for other page(s)

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    [Inject] public HttpClient HttpClient { get; set; } = default!;
+
+    [Inject] public NavigationManager NavigationManager { get; set; } = default!;
+
+    [Inject] public ITodoTemplateAuthenticationService TodoTemplateAuthenticationService { get; set; } = default!;
+
+    public BitMessageBarType ResetPasswordMessageType { get; set; }
+    public string? ResetPasswordMessage { get; set; }
+
+    protected override void OnInitialized()
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:31]: shouldn't we use the `OnInitAsync` method instead?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs (hunk 9 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:35]: again, url-encoding should be better.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignIn.razor [hunk 12]
```diff
@@ -19,7 +19,7 @@
             <BitLabel>Email</BitLabel>
             <BitTextField @bind-Value="Email"
                 Placeholder="Email"
-                OnKeyUp="WrapHandled(CheckSignInButtonEnabled)"
+                OnChange="WrapHandled(CheckSignInButtonEnabled)"
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignIn.razor:22]: plz do not submit any changes that are not related to the issue

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignIn.razor [hunk 13]
```diff
@@ -29,7 +29,7 @@
                 Type="BitTextFieldType.Password" 
                 CanRevealPassword="true"
                 Placeholder="Password"
-                OnKeyUp="WrapHandled(CheckSignInButtonEnabled)"
+                OnChange="WrapHandled(CheckSignInButtonEnabled)"
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignIn.razor:32]: plz do not submit any changes that are not related to the issue

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor [hunk 14]
```diff
@@ -21,7 +21,7 @@
             <BitLabel>Email</BitLabel>
             <BitTextField @bind-Value="Email"
                           Placeholder="Email"
-                          OnKeyUp="WrapHandled(CheckSignUpButtonEnable)"
+                          OnChange="WrapHandled(CheckSignUpButtonEnable)"
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor:24]: plz do not submit any changes that are not related to the issue

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor [hunk 15]
```diff
@@ -30,7 +30,7 @@
             <BitTextField @bind-Value="Password"
                           Type="BitTextFieldType.Password"
                           CanRevealPassword="true" Placeholder="Password"
-                          OnKeyUp="WrapHandled(CheckSignUpButtonEnable)"
+                          OnChange="WrapHandled(CheckSignUpButtonEnable)"
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor:33]: plz do not submit any changes that are not related to the issue

**Review by msynk — CHANGES_REQUESTED** [review]

**Comment by msynk** [comment]
please do not resolve any comment before submitting the required changes

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs (hunk 11 above)
Comment (ysmoradi) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:31]: We need another issue to make OnInitialized sealed so no one can override it accidentally @msynk @RezaKargar

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor (hunk 7 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/EditProfile.razor:61]: It handles all scenarios, however, I removed it from the PR, and I'll follow it with @maryamhdr. @msynk @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: > What's the difference between the reset password token and email confirmation token? I think both need or don't need URL encoding? @mohabbati

In the email confirmation, we don't have token so there is no illegal character, but in the reset password, we have.

Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: I just encoded token using HttpUtility.UrlEncode() in the link generation. The point is that in the razor partial class it does not need to decode. It seems asp .net decodes it automatically. Is it enough? @msynk @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs (hunk 9 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:35]: Duplicate.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (ysmoradi) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: Maybe, what do you think? @msynk 
Meanwhile, could you please test your theory with an illegal character? @mohabbati

Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:114]: I tested it with illegal characters and it worked well. @ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs (hunk 11 above)
Comment (mohabbati) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ResetPassword.razor.cs:31]: Due to auto decode, I removed OnInitialized() method. @ysmoradi @msynk

**Review by ysmoradi — APPROVED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs (hunk 8 above)
Comment (RezaKargar) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/ForgotPassword.razor.cs:61]: > @RezaKargar Look at this, I think you'd be interested.

@ysmoradi  I just created an issue (#1747) for it, check it out please

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs (hunk 4 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs]: yeah, the decoding part is done automatically by the framework itself.
great job guys :)

**Review by msynk — APPROVED** [review]
